docs/_guide/events.md

```diff
@@ -8,96 +8,156 @@ slug: events
-    
-    *   Removing the event listener in `disconnectedCallback` ensures that any memory allocated by your component will be cleaned up when your component is destroyed or disconnected from the page. 
+**Example**
 
-    *   Adding the event listener in `connectedCallback` (instead of, for example, the constructor or `firstRendered`) ensures that your component will re-create its event listener if it is disconnected and subsequently reconnected to DOM.
-    
-    **Example**
+```js
+connectedCallback() {
+  super.connectedCallback();
+  window.addEventListener('resize', this.handleResize);
+}
+disconnectedCallback() {
+  window.removeEventListener('resize', this.handleResize);
+  super.disconnectedCallback();
+}
+```
+
+### Add event listeners after first paint
+
+Sometimes, you may want to defer adding an event listener until after first paint—for example, if you're adding a lot of listeners and first paint performance is critical.
+
+LitElement doesn't have a specific lifecycle callback called after first paint, but you can use this pattern with the `firstUpdated` lifecycle callback:
+
+```js
+async firstUpdated() {
+  // Give the browser a chance to paint
+  await new Promise((r) => setTimeout(r, 0));
+  this.addEventListener('click', this._handleClick);
+}
+```
+
+`firstUpdated` fires after the first time your component has been updated and called its `render` method, but **before the browser has had a chance to paint**. The `Promise`/`setTimeout` line yields to the browser
     
-    ```js
-    connectedCallback() {
-      super.connectedCallback();
-      document.addEventListener('readystatechange', this.handleChange);
-    }
-    disconnectedCallback() {
-      document.removeEventListener('readystatechange', this.handleChange);
-      super.disconnectedCallback();
-    }
-    ```
+See [firstUpdated](/guide/lifecycle#firstupdated) in the Lifecycle documentation for more information.
+
 
-### Using `this` in event handlers
+## Using `this` in event listeners
 
-The default JavaScript context object (`this`) inside an event handler belonging to a LitElement-based component is the component itself. 
+Event listeners added using the declarative (`@event`) syntax in the template are automatically _bound_ to the component.
 
-Therefore, you can use `this` to refer to your element instance inside any event handler:
+Therefore, you can use `this` to refer to your component instance inside any declarative event handler:
 
 ```js
 class MyElement extends LitElement {
   render() {
-    return html`<button @click="${this.handleClick}">click</button>`;
+    return html`<button @click="${this._handleClick}">click</button>`;
   }
-  handleClick(e) {
+  _handleClick(e) {
     console.log(this.prop);
   }
 }
 ```
 
+When adding listeners imperatively with `addEventListener`, you'll need to bind the event listener yourself if you need a reference to the component instance. For example:
+
+```js
+this.boundResizeHandler = this.handleResize.bind(this);
```

Comment:
I think the most ergonomic way to do this is like:

```ts
class MyElement extends LitElement {
  private _onResize = (e) => { ... };

  constructor() {
    super();
    window.addEventListener('resize', this._onResize);
  }
}
```

The only downside here is you can't call `super._onResize()` in the handler, but that's a rare thing to need to do.

Reply:
There is also a certain weird tax. I've seen this before and I still had to read a half-dozen pages of doc to remind myself why it works.

Also: assumes class fields support. Should I include both examples, perhaps?

Reply:
Included both.